### PR TITLE
SDK-390: message deserialize checks the magic bytes right after havin…

### PR DESCRIPTION
**Pull request by paologalligit**
…g it extracted

### CODE REVIEW

File: src/main/scala/sparkz/core/network/message/MessageSerializer.scala [hunk 1]
```diff
@@ -40,6 +40,10 @@ class MessageSerializer(specs: Seq[MessageSpec[_]], magicBytes: Array[Byte], mes
     } else {
       val it = byteString.iterator
       val magic = it.getBytes(MagicLength)
+      //peer is from another network
+      if (!java.util.Arrays.equals(magic, magicBytes)) {
+        throw MaliciousBehaviorException(s"Wrong magic bytes, expected ${magicBytes.mkString}, got ${magic.mkString} in : ${byteString.utf8String}")
+      }
       val msgCode = it.getByte
```
Comment (i-Alex) [inline src/main/scala/sparkz/core/network/message/MessageSerializer.scala:46]: Let's check more careful the case, when we retrieved the message with the code which node spec registered:
`val spec = specsMap.getOrElse(msgCode, throw new Error(s"No message handler found for $msgCode"))`.

Here I can see that:
1) we may check the spec right now, no need to wait for the full message.
2) will we stuck reading the same message? I don't see the code that removes this data from `chunksBuffer` in `PeerConnectionHandler.scala`. Check if we already have some UTs for this method. If not - let's add  a couple to confirm that it works/doesn't work

**Review by i-Alex — CHANGES_REQUESTED** [review]
The original feature seems good.
Another point appeared.

File: src/main/scala/sparkz/core/network/PeerConnectionHandler.scala [hunk 2]
```diff
@@ -204,7 +204,7 @@ class PeerConnectionHandler(val settings: NetworkSettings,
               case MaliciousBehaviorException(msg) =>
                 log.warn(s"Banning peer for malicious behaviour($msg): ${connectionId.toString}")
                 //peer will be added to the blacklist and the network controller will send CloseConnection
```
Comment (i-skrypnyk) [inline src/main/scala/sparkz/core/network/PeerConnectionHandler.scala:206]: this comment seems outdated now

File: src/main/scala/sparkz/core/network/peer/InMemoryPeerDatabase.scala [hunk 3]
```diff
@@ -119,11 +119,11 @@ final class InMemoryPeerDatabase(settings: NetworkSettings, timeProvider: TimePr
         case Some((penaltyScoreAcc, lastPenaltyTs)) =>
           val currentTime = timeProvider.time()
           if (currentTime - lastPenaltyTs - safeInterval > 0 || penaltyType.isPermanent)
-            (penaltyScoreAcc + penaltyScore(penaltyType), timeProvider.time())
+            (penaltyScoreAcc + penaltyType.penaltyScore, timeProvider.time())
```
Comment (i-skrypnyk) [inline src/main/scala/sparkz/core/network/peer/InMemoryPeerDatabase.scala:122]: :fire: :+1:

**Review by i-skrypnyk — APPROVED** [review]
LGTM

File: src/main/scala/sparkz/core/network/PeerConnectionHandler.scala (hunk 2 above)
Comment (i-Alex) [reply src/main/scala/sparkz/core/network/PeerConnectionHandler.scala:206]: seems ok after the last changes with disconnectPenalty.
@paologalligit pls check it

**Review by i-Alex — APPROVED** [review]

File: src/main/scala/sparkz/core/network/PeerConnectionHandler.scala (hunk 2 above)
Comment (paologalligit) [reply src/main/scala/sparkz/core/network/PeerConnectionHandler.scala:206]: actually, Ivan was quite right: although the intention was to ban the peer with the DisconnectPenalty temporarily, the implementation was bugged. To penalize the peer we need the penalty score to go over the threshold, while the DisconnectPenalty is reaching the exact score of the threshold resulting in a penalization but not a ban. Pushing the fix for it; thanks

**Pull request merged**